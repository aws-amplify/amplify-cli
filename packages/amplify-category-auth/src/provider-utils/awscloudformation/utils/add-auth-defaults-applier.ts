import { ServiceQuestionsResult } from '../service-walkthrough-types';
import { verificationBucketName } from './verification-bucket-name';
import { merge } from 'lodash';

/**
 * Factory function that returns a function that applies default values to a ServiceQuestionsResult request.
 * It does not overwrite existing values in the request.
 *
 * The logic here has been refactored from service-walkthroughs/auth-questions.js and is mostly unchanged
 * @param defaultValuesFilename The filename to fetch defaults from
 * @param projectName The name of the current project (used to generate some default values)
 */
export const getAddAuthDefaultsApplier = (defaultValuesFilename: string, projectName: string) => async (
  result: ServiceQuestionsResult,
): Promise<ServiceQuestionsResult> => {
  const { functionMap, generalDefaults, roles } = require(`../assets/${defaultValuesFilename}`);
  result = merge(generalDefaults(projectName), result);

  await verificationBucketName(result);

  /* merge actual answers object into props object,
   * ensuring that manual entries override defaults */
  return merge(functionMap[result.authSelections](result.resourceName), result, roles);
};
