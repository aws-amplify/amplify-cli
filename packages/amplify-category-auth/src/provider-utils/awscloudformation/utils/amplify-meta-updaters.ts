import * as path from 'path';
import { JSONUtilities, pathManager, $TSAny } from 'amplify-cli-core';
import { hostedUIProviders } from '../assets/string-maps';
import { AuthParameters } from '../import/types';

/**
 * Factory function that returns a function that updates Amplify meta files after adding auth resource assets
 *
 * refactored from commands/enable.js
 * @param context The amplify context
 * @param resultMetadata The metadata from the service selection prompt
 */
export const getPostAddAuthMetaUpdater =
  (context: any, resultMetadata: { service: string; providerName: string }) =>
  (resourceName: string): string => {
    const options: any = {
      service: resultMetadata.service,
      providerPlugin: resultMetadata.providerName,
    };
    const parametersJSONPath = path.join(context.amplify.pathManager.getBackendDirPath(), 'auth', resourceName, 'build', 'parameters.json');
    const authParameters = JSONUtilities.readJson<AuthParameters>(parametersJSONPath)!;

    if (authParameters.dependsOn) {
      options.dependsOn = authParameters.dependsOn;
    }

    let customAuthConfigured = false;
    if (authParameters.triggers) {
      const triggers = JSONUtilities.parse<any>(authParameters.triggers);

      customAuthConfigured =
        !!triggers.DefineAuthChallenge &&
        triggers.DefineAuthChallenge.length > 0 &&
        !!triggers.CreateAuthChallenge &&
        triggers.CreateAuthChallenge.length > 0 &&
        !!triggers.VerifyAuthChallengeResponse &&
        triggers.VerifyAuthChallengeResponse.length > 0;
    }

    options.customAuth = customAuthConfigured;
    options.frontendAuthConfig = getFrontendConfig(authParameters);

    context.amplify.updateamplifyMetaAfterResourceAdd('auth', resourceName, options);

    // Remove Identity Pool dependency attributes on userpool groups if Identity Pool not enabled
    const allResources = context.amplify.getProjectMeta();
    if (allResources.auth && allResources.auth.userPoolGroups) {
      if (!authParameters.identityPoolName) {
        const userPoolGroupDependsOn = [
          {
            category: 'auth',
            resourceName,
            attributes: ['UserPoolId', 'AppClientIDWeb', 'AppClientID'],
          },
        ];
        context.amplify.updateamplifyMetaAfterResourceUpdate('auth', 'userPoolGroups', 'dependsOn', userPoolGroupDependsOn);
      }
    }
    return resourceName;
  };

/**
 * Factory function that returns a function that updates Amplify meta files after updating auth resource assets
 * @param context The amplify context
 */
export const getPostUpdateAuthMetaUpdater = (context: any) => async (resourceName: string) => {
  const resourceDirPath = path.join(pathManager.getBackendDirPath(), 'auth', resourceName, 'build', 'parameters.json');
  const authParameters = JSONUtilities.readJson<AuthParameters>(resourceDirPath)!;
  if (authParameters.dependsOn) {
    context.amplify.updateamplifyMetaAfterResourceUpdate('auth', resourceName, 'dependsOn', authParameters.dependsOn);
  }

  let customAuthConfigured = false;
  if (authParameters.triggers) {
    const triggers = JSONUtilities.parse<$TSAny>(authParameters.triggers);
    customAuthConfigured =
      !!triggers.DefineAuthChallenge &&
      triggers.DefineAuthChallenge.length > 0 &&
      !!triggers.CreateAuthChallenge &&
      triggers.CreateAuthChallenge.length > 0 &&
      !!triggers.VerifyAuthChallengeResponse &&
      triggers.VerifyAuthChallengeResponse.length > 0;
  }
  context.amplify.updateamplifyMetaAfterResourceUpdate('auth', resourceName, 'customAuth', customAuthConfigured);
  context.amplify.updateamplifyMetaAfterResourceUpdate('auth', resourceName, 'frontendAuthConfig', getFrontendConfig(authParameters));

  // Update Identity Pool dependency attributes on userpool groups
  const allResources = context.amplify.getProjectMeta();
  if (allResources.auth && allResources.auth.userPoolGroups) {
    let attributes = ['UserPoolId', 'AppClientIDWeb', 'AppClientID'];
    if (authParameters.identityPoolName) {
      attributes.push('IdentityPoolId');
    }
    const userPoolGroupDependsOn = [
      {
        category: 'auth',
        resourceName,
        attributes,
      },
    ];

    context.amplify.updateamplifyMetaAfterResourceUpdate('auth', 'userPoolGroups', 'dependsOn', userPoolGroupDependsOn);
  }
  return resourceName;
};

export function getFrontendConfig(authParameters: AuthParameters) {
  const verificationMechanisms = (authParameters?.autoVerifiedAttributes || []).map((att: string) => att.toUpperCase());
  const usernameAttributes: string[] = [];

  if (authParameters?.usernameAttributes && authParameters.usernameAttributes.length > 0) {
    authParameters.usernameAttributes[0].split(',').forEach(it => usernameAttributes.push(it.trim().toUpperCase()));
  }

  const socialProviders: string[] = [];
  (authParameters?.authProvidersUserPool ?? []).forEach((provider: string) => {
    const key = hostedUIProviders.find(it => it.value === provider)?.key;

    if (key) {
      socialProviders.push(key);
    }
  });

  const signupAttributes = (authParameters?.requiredAttributes || []).map((att: string) => att.toUpperCase());

  const passwordProtectionSettings = {
    passwordPolicyMinLength: authParameters?.passwordPolicyMinLength,
    passwordPolicyCharacters: (authParameters?.passwordPolicyCharacters || []).map((i: string) => i.replace(/ /g, '_').toUpperCase()),
  };

  const mfaTypes: string[] = [];
  if (authParameters.mfaTypes) {
    if (authParameters.mfaTypes.includes('SMS Text Message')) {
      mfaTypes.push('SMS');
    }

    if (authParameters.mfaTypes.includes('TOTP')) {
      mfaTypes.push('TOTP');
    }
  }

  return {
    socialProviders: socialProviders,
    usernameAttributes: usernameAttributes,
    signupAttributes: signupAttributes,
    passwordProtectionSettings: passwordProtectionSettings,
    mfaConfiguration: authParameters?.mfaConfiguration,
    mfaTypes: mfaTypes,
    verificationMechanisms: verificationMechanisms,
  };
}
