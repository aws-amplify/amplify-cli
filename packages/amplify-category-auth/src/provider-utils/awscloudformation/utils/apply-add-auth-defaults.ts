import { ServiceQuestionsResult } from '../legacy-types';
import { verificationBucketName } from './verification-bucket-name';
import { merge } from 'lodash';

export const getAddAuthDefaultsApplier = (defaultValuesFilename: string, projectName: string) => async (
  result: ServiceQuestionsResult,
): Promise<ServiceQuestionsResult> => {
  const { functionMap, generalDefaults, roles } = require(`../assets/${defaultValuesFilename}`);
  result = merge(generalDefaults(projectName), result);

  await verificationBucketName(result);

  /* merge actual answers object into props object,
   * ensuring that manual entries override defaults */
  return merge(functionMap[result.authSelections](result.resourceName), result, roles);
};
