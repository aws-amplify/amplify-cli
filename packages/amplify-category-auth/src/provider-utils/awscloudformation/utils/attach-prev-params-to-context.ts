import { getAuthResourceName } from '../../../utils/getAuthResourceName';
import { AuthInputState } from '../auth-inputs-manager/auth-input-state';

/**
 * Some existing code relies on the existing auth parameters being attached to context.updatingAuth
 */
export const attachPrevParamsToContext = async (context: any) => {
  const resourceName = await getAuthResourceName(context);
  const cliState = new AuthInputState(resourceName);
  context.updatingAuth = await cliState.loadResourceParameters(context, cliState.getCLIInputPayload());
};
