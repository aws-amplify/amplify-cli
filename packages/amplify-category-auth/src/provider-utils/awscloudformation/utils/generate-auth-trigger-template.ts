import * as path from 'path';
import * as fs from 'fs-extra';
import { $TSAny, AmplifyFault, JSONUtilities, pathManager, getPermissionsBoundaryArn } from '@aws-amplify/amplify-cli-core';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import { CustomResource } from 'aws-cdk-lib';
import { v4 as uuid } from 'uuid';
import { Construct } from 'constructs';
import { authTriggerAssetFilePath } from '../constants';
import _ from 'lodash';
import { AuthTriggerConnection, AuthTriggerPermissions, CognitoStackOptions } from '../service-walkthrough-types/cognito-user-input-types';
import { configureSmsOption } from './configure-sms';

type CustomResourceAuthStackProps = Readonly<{
  description: string;
  authTriggerConnections: AuthTriggerConnection[];
  enableSnsRole: boolean;
  permissions?: AuthTriggerPermissions[];
}>;

const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';

/**
 * CDK stack for custom auth resources
 */
export class CustomResourceAuthStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CustomResourceAuthStackProps) {
    super(scope, id, { ...props, synthesizer: new cdk.LegacyStackSynthesizer() });
    this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;

    const env = new cdk.CfnParameter(this, 'env', {
      type: 'String',
    });

    const userpoolId = new cdk.CfnParameter(this, 'userpoolId', {
      type: 'String',
    });

    const userpoolArn = new cdk.CfnParameter(this, 'userpoolArn', {
      type: 'String',
    });

    // eslint-disable-next-line no-new
    new cdk.CfnCondition(this, 'ShouldNotCreateEnvResources', {
      expression: cdk.Fn.conditionEquals(env, 'NONE'),
    });

    props.authTriggerConnections.forEach((triggerConfig) => {
      const config = triggerConfig;
      const fnName = new cdk.CfnParameter(this, `function${config.lambdaFunctionName}Name`, {
        type: 'String',
      });
      const fnArn = new cdk.CfnParameter(this, `function${config.lambdaFunctionName}Arn`, {
        type: 'String',
      });
      createPermissionToInvokeLambda(this, fnName, userpoolArn, config);
      const roleArn = new cdk.CfnParameter(this, `function${config.lambdaFunctionName}LambdaExecutionRole`, {
        type: 'String',
      });
      config.lambdaFunctionArn = fnArn.valueAsString;

      if (!_.isEmpty(props.permissions)) {
        const lambdaPermission = props.permissions!.find((permission) => config.triggerType === permission.trigger);
        if (!_.isEmpty(lambdaPermission)) {
          createPermissionsForAuthTrigger(this, fnName, roleArn, lambdaPermission!, userpoolArn);
        }
      }
    });

    createCustomResource(this, props.authTriggerConnections, userpoolId, userpoolArn, props.enableSnsRole);
  }

  /**
   * This function renderers a full CFN template for this stack.
   * It is inspired by
   * https://github.com/aws/aws-cdk/blob/bd056d1d38a2d3f43efe4f857c4d38b30fb9b681/packages/%40aws-cdk/assertions/lib/template.ts#L298-L310.
   * This replaces private prepareApp (from CDK v1) and this._toCloudFormation() (the latter does not function properly without the former).
   */
  toCloudFormation = (): $TSAny => {
    const root = this.node.root as cdk.Stage;
    const assembly = root.synth();
    if (!this.nestedStackParent) {
      return assembly.getStackArtifact(this.artifactId).template;
    }
    // if this is a nested stack ( i.e. it has a parent), then just read the template as a string
    const template = fs.readFileSync(path.join(assembly.directory, this.templateFile));
    return JSON.parse(template.toString('utf-8'));
  };
}

/**
 * Creates nested auth trigger CFN template and writes it to the project directory
 */
export const generateNestedAuthTriggerTemplate = async (
  category: string,
  resourceName: string,
  request: CognitoStackOptions,
): Promise<void> => {
  const cfnFileName = 'auth-trigger-cloudformation-template.json';
  const targetDir = path.join(pathManager.getBackendDirPath(), category, resourceName, 'build');
  const authTriggerCfnFilePath = path.join(targetDir, cfnFileName);
  const { authTriggerConnections, permissions, useEnabledMfas } = request;

  const configureSMS = configureSmsOption(request);

  const enableSnsRole: boolean | undefined = !useEnabledMfas || configureSMS;

  if (!_.isEmpty(authTriggerConnections)) {
    const cfnObject = await createCustomResourceForAuthTrigger(authTriggerConnections!, !!enableSnsRole, permissions);
    JSONUtilities.writeJson(authTriggerCfnFilePath, cfnObject);
  } else {
    // delete the custom stack template if the triggers aren't defined
    try {
      fs.unlinkSync(authTriggerCfnFilePath);
    } catch (err) {
      // if its not present do nothing
    }
  }
};

/**
 * creates custom resource for cognito triggers
 */
// eslint-disable-next-line spellcheck/spell-checker
export const createCustomResourceForAuthTrigger = async (
  authTriggerConnections: AuthTriggerConnection[],
  enableSnsRole: boolean,
  permissions?: AuthTriggerPermissions[],
): Promise<$TSAny> => {
  if (Array.isArray(authTriggerConnections) && authTriggerConnections.length) {
    const stack = new CustomResourceAuthStack(undefined as $TSAny, 'Amplify', {
      description: 'Custom Resource stack for Auth Trigger created using Amplify CLI',
      authTriggerConnections,
      enableSnsRole,
      permissions,
    });
    const cfn = stack.toCloudFormation();
    return cfn;
  }
  throw new AmplifyFault('AuthCategoryFault', {
    message: `Auth Trigger Connections must have value when trigger are selected`,
  });
};

const createCustomResource = (
  stack: cdk.Stack,
  authTriggerConnections: AuthTriggerConnection[],
  userpoolId: cdk.CfnParameter,
  userpoolArn: cdk.CfnParameter,
  enableSnsRole: boolean,
): void => {
  const triggerCode = fs.readFileSync(authTriggerAssetFilePath, 'utf-8');
  const authTriggerFn = new lambda.Function(stack, 'authTriggerFn', {
    runtime: lambda.Runtime.NODEJS_18_X,
    code: lambda.Code.fromInline(triggerCode),
    handler: 'index.handler',
  });

  if (authTriggerFn.role) {
    authTriggerFn.role.addToPrincipalPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['cognito-idp:DescribeUserPool', 'cognito-idp:UpdateUserPool'],
        resources: [userpoolArn.valueAsString],
      }),
    );

    const policyArn = getPermissionsBoundaryArn();
    if (policyArn) {
      iam.PermissionsBoundary.of(authTriggerFn).apply(iam.ManagedPolicy.fromManagedPolicyArn(stack, 'PermissionsBoundary', policyArn));
    }

    // reason to add iam::PassRole
    // AccessDeniedException: User: <IAM User> is not authorized to perform: iam:PassRole
    // on resource: <auth trigger role>  if (authTriggerFn.role) {
    if (enableSnsRole) {
      const snsRoleArn = new cdk.CfnParameter(stack, 'snsRoleArn', {
        type: 'String',
      });
      authTriggerFn.role.addToPrincipalPolicy(
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['iam:PassRole'],
          resources: [snsRoleArn.valueAsString],
        }),
      );
    }
  }

  // The custom resource that uses the provider to supply value
  // Passing in a nonce parameter to ensure that the custom resource is triggered on every deployment
  // eslint-disable-next-line no-new
  const customResource = new CustomResource(stack, 'CustomAuthTriggerResource', {
    serviceToken: authTriggerFn.functionArn,
    properties: { userpoolId: userpoolId.valueAsString, lambdaConfig: authTriggerConnections, nonce: uuid() },
    resourceType: 'Custom::CustomAuthTriggerResourceOutputs',
  });

  customResource.node.addDependency(authTriggerFn);
};

const createPermissionToInvokeLambda = (
  stack: cdk.Stack,
  fnName: cdk.CfnParameter,
  userpoolArn: cdk.CfnParameter,
  config: AuthTriggerConnection,
): void => {
  // eslint-disable-next-line no-new
  new lambda.CfnPermission(stack, `UserPool${config.triggerType}LambdaInvokePermission`, {
    action: 'lambda:InvokeFunction',
    functionName: fnName.valueAsString,
    principal: 'cognito-idp.amazonaws.com',
    sourceArn: userpoolArn.valueAsString,
  });
};

const createPermissionsForAuthTrigger = (
  stack: cdk.Stack,
  fnName: cdk.CfnParameter,
  roleArn: cdk.CfnParameter,
  permissions: AuthTriggerPermissions,
  userpoolArn: cdk.CfnParameter,
): iam.Policy => {
  const myRole = iam.Role.fromRoleArn(stack, 'LambdaExecutionRole', roleArn.valueAsString);
  return new iam.Policy(stack, `${fnName}${permissions.trigger}${permissions.policyName}`, {
    policyName: permissions.policyName,
    statements: [
      new iam.PolicyStatement({
        effect: permissions.effect === iam.Effect.ALLOW ? iam.Effect.ALLOW : iam.Effect.DENY,
        actions: permissions.actions,
        resources: [userpoolArn.valueAsString],
      }),
    ],
    roles: [myRole],
  });
};
