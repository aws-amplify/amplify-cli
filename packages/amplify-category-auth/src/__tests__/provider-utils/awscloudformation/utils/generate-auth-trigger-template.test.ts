import { getPermissionsBoundaryArn } from '@aws-amplify/amplify-cli-core';
import {
  AuthTriggerConnection,
  AuthTriggerPermissions,
  TriggerType,
} from '../../../../provider-utils/awscloudformation/service-walkthrough-types/cognito-user-input-types';
import { createCustomResourceForAuthTrigger } from '../../../../provider-utils/awscloudformation/utils/generate-auth-trigger-template';
jest.mock('@aws-amplify/amplify-cli-core');
jest.mock('uuid');
describe('generate Auth Trigger Template', () => {
  it('successfully generate IAM policies when permissions are defined', async () => {
    const mockAuthTriggerConnections: AuthTriggerConnection[] = [
      {
        lambdaFunctionName: 'randomFn',
        triggerType: TriggerType.PostConfirmation,
        lambdaFunctionArn: 'randomArn',
      },
    ];

    const mockAuthTriggerPermissions: AuthTriggerPermissions[] = [
      {
        policyName: 'AddToGroupCognito',
        trigger: 'PostConfirmation',
        effect: 'Allow',
        actions: ['cognito-idp:AdminAddUserToGroup', 'cognito-idp:GetGroup', 'cognito-idp:CreateGroup'],
        resource: {
          paramType: '!GetAtt',
          keys: ['UserPool', 'Arn'],
        },
      },
    ];
    const cfn = await createCustomResourceForAuthTrigger(mockAuthTriggerConnections, false, mockAuthTriggerPermissions);
    expect(cfn.Resources.AmplifyfunctionrandomFnNamePostConfirmationAddToGroupCognito1120888F).toMatchInlineSnapshot(`
{
  "Properties": {
    "PolicyDocument": {
      "Statement": [
        {
          "Action": [
            "cognito-idp:AdminAddUserToGroup",
            "cognito-idp:GetGroup",
            "cognito-idp:CreateGroup",
          ],
          "Effect": "Allow",
          "Resource": {
            "Ref": "userpoolArn",
          },
        },
      ],
      "Version": "2012-10-17",
    },
    "PolicyName": "AddToGroupCognito",
    "Roles": [
      {
        "Fn::Select": [
          1,
          {
            "Fn::Split": [
              "/",
              {
                "Fn::Select": [
                  5,
                  {
                    "Fn::Split": [
                      ":",
                      {
                        "Ref": "functionrandomFnLambdaExecutionRole",
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  "Type": "AWS::IAM::Policy",
}
`);
    // No permissions boundary set if not available
    expect(cfn.Resources.authTriggerFnServiceRole08093B67.Properties.PermissionsBoundary).toBeUndefined();
  });

  it('adds permissions boundary to new role if available', async () => {
    // Setup for a successfully generating custom resource.
    const mockAuthTriggerConnections: AuthTriggerConnection[] = [
      {
        lambdaFunctionName: 'randomFn',
        triggerType: TriggerType.PostConfirmation,
        lambdaFunctionArn: 'randomArn',
      },
    ];
    const mockAuthTriggerPermissions: AuthTriggerPermissions[] = [
      {
        policyName: 'AddToGroupCognito',
        trigger: 'PostConfirmation',
        effect: 'Allow',
        actions: ['cognito-idp:AdminAddUserToGroup', 'cognito-idp:GetGroup', 'cognito-idp:CreateGroup'],
        resource: {
          paramType: '!GetAtt',
          keys: ['UserPool', 'Arn'],
        },
      },
    ];

    // ensure permissions boundary is set
    const getPermissionsBoundaryArn_mock = getPermissionsBoundaryArn as jest.MockedFunction<typeof getPermissionsBoundaryArn>;
    getPermissionsBoundaryArn_mock.mockReturnValue('testPermissionsBoundaryArn');

    // execute the test
    const cfn = await createCustomResourceForAuthTrigger(mockAuthTriggerConnections, false, mockAuthTriggerPermissions);

    //validate
    expect(cfn.Resources.authTriggerFnServiceRole08093B67.Properties.PermissionsBoundary).toStrictEqual('testPermissionsBoundaryArn');

    // Restore for other folks
    getPermissionsBoundaryArn_mock.mockRestore();
  });

  it('does not generate iam policies for Auth trigger when permissions are empty', async () => {
    const mockAuthTriggerConnections: AuthTriggerConnection[] = [
      {
        lambdaFunctionName: 'randomFn',
        triggerType: TriggerType.PreAuthentication,
        lambdaFunctionArn: 'randomArn',
      },
    ];

    const mockAuthTriggerPermissions: AuthTriggerPermissions[] = [];
    // eslint-disable-next-line spellcheck/spell-checker
    const cfn = await createCustomResourceForAuthTrigger(mockAuthTriggerConnections, false, mockAuthTriggerPermissions);
    expect(cfn.Resources.AmplifyfunctionrandomFnNamePostConfirmationAddToGroupCognito1120888F).not.toBeDefined();
  });
});

describe('generateNestedAuthTriggerTemplate', () => {
  it('adds "authTriggerFn" as a dependency on "CustomAuthTriggerResource"', async () => {
    const authTriggerConnections = [
      {
        triggerType: 'CustomMessage',
        lambdaFunctionName: 'authtestCustomMessage',
      },
      {
        triggerType: 'PostConfirmation',
        lambdaFunctionName: 'authtestostConfirmation',
      },
    ];

    const cfnTemplate = await createCustomResourceForAuthTrigger(authTriggerConnections, false);

    expect(cfnTemplate).toMatchSnapshot();
    expect(cfnTemplate.Resources.CustomAuthTriggerResource.DependsOn).toEqual(
      expect.arrayContaining([
        'authTriggerFn7FCFA449',
        'authTriggerFnServiceRoleDefaultPolicyEC9285A8',
        'authTriggerFnServiceRole08093B67',
      ]),
    );
  });
});
