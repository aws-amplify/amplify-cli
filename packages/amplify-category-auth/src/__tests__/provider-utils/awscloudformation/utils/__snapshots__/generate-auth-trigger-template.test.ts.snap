// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateNestedAuthTriggerTemplate adds "authTriggerFn" as a dependency on "CustomAuthTriggerResource" 1`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Conditions": {
    "ShouldNotCreateEnvResources": {
      "Fn::Equals": [
        {
          "Ref": "env",
        },
        "NONE",
      ],
    },
  },
  "Description": "Custom Resource stack for Auth Trigger created using Amplify CLI",
  "Parameters": {
    "env": {
      "Type": "String",
    },
    "functionauthtestCustomMessageArn": {
      "Type": "String",
    },
    "functionauthtestCustomMessageLambdaExecutionRole": {
      "Type": "String",
    },
    "functionauthtestCustomMessageName": {
      "Type": "String",
    },
    "functionauthtestostConfirmationArn": {
      "Type": "String",
    },
    "functionauthtestostConfirmationLambdaExecutionRole": {
      "Type": "String",
    },
    "functionauthtestostConfirmationName": {
      "Type": "String",
    },
    "userpoolArn": {
      "Type": "String",
    },
    "userpoolId": {
      "Type": "String",
    },
  },
  "Resources": {
    "CustomAuthTriggerResource": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "authTriggerFn7FCFA449",
        "authTriggerFnServiceRoleDefaultPolicyEC9285A8",
        "authTriggerFnServiceRole08093B67",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "authTriggerFn7FCFA449",
            "Arn",
          ],
        },
        "lambdaConfig": [
          {
            "lambdaFunctionArn": {
              "Ref": "functionauthtestCustomMessageArn",
            },
            "lambdaFunctionName": "authtestCustomMessage",
            "triggerType": "CustomMessage",
          },
          {
            "lambdaFunctionArn": {
              "Ref": "functionauthtestostConfirmationArn",
            },
            "lambdaFunctionName": "authtestostConfirmation",
            "triggerType": "PostConfirmation",
          },
        ],
        "userpoolId": {
          "Ref": "userpoolId",
        },
      },
      "Type": "Custom::CustomAuthTriggerResourceOutputs",
      "UpdateReplacePolicy": "Delete",
    },
    "UserPoolCustomMessageLambdaInvokePermission": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Ref": "functionauthtestCustomMessageName",
        },
        "Principal": "cognito-idp.amazonaws.com",
        "SourceArn": {
          "Ref": "userpoolArn",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "UserPoolPostConfirmationLambdaInvokePermission": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Ref": "functionauthtestostConfirmationName",
        },
        "Principal": "cognito-idp.amazonaws.com",
        "SourceArn": {
          "Ref": "userpoolArn",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "authTriggerFn7FCFA449": {
      "DependsOn": [
        "authTriggerFnServiceRoleDefaultPolicyEC9285A8",
        "authTriggerFnServiceRole08093B67",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "const response = require('cfn-response');
const {
  CognitoIdentityProviderClient,
  DescribeUserPoolCommand,
  UpdateUserPoolCommand,
} = require('@aws-sdk/client-cognito-identity-provider');

exports.handler = (event, context) => {
  // Don't return promise, response.send() marks context as done internally
  void tryHandleEvent(event, context);
};

async function tryHandleEvent(event, context) {
  const physicalResourceId =
    event.RequestType === 'Update' ? event.PhysicalResourceId : \`\${event.LogicalResourceId}-\${event.ResourceProperties.userpoolId}\`;
  try {
    await handleEvent(event);
    response.send(event, context, response.SUCCESS, {}, physicalResourceId);
  } catch (err) {
    console.log(err.stack);
    response.send(event, context, response.FAILED, { err }, physicalResourceId);
  }
}

async function handleEvent(event) {
  const userPoolId = event.ResourceProperties.userpoolId;
  const { lambdaConfig } = event.ResourceProperties;
  const config = {};
  const cognitoClient = new CognitoIdentityProviderClient({});
  const userPoolConfig = await cognitoClient.send(new DescribeUserPoolCommand({ UserPoolId: userPoolId }));
  const userPoolParams = userPoolConfig.UserPool;
  // update userPool params

  const updateUserPoolConfig = {
    UserPoolId: userPoolParams.Id,
    Policies: userPoolParams.Policies,
    SmsVerificationMessage: userPoolParams.SmsVerificationMessage,
    AccountRecoverySetting: userPoolParams.AccountRecoverySetting,
    AdminCreateUserConfig: userPoolParams.AdminCreateUserConfig,
    AutoVerifiedAttributes: userPoolParams.AutoVerifiedAttributes,
    EmailConfiguration: userPoolParams.EmailConfiguration,
    EmailVerificationMessage: userPoolParams.EmailVerificationMessage,
    EmailVerificationSubject: userPoolParams.EmailVerificationSubject,
    VerificationMessageTemplate: userPoolParams.VerificationMessageTemplate,
    SmsAuthenticationMessage: userPoolParams.SmsAuthenticationMessage,
    MfaConfiguration: userPoolParams.MfaConfiguration,
    DeviceConfiguration: userPoolParams.DeviceConfiguration,
    SmsConfiguration: userPoolParams.SmsConfiguration,
    UserPoolTags: userPoolParams.UserPoolTags,
    UserPoolAddOns: userPoolParams.UserPoolAddOns,
  };

  // removing undefined keys
  Object.keys(updateUserPoolConfig).forEach((key) => updateUserPoolConfig[key] === undefined && delete updateUserPoolConfig[key]);

  /* removing UnusedAccountValidityDays as deprecated
    InvalidParameterException: Please use TemporaryPasswordValidityDays in PasswordPolicy instead of UnusedAccountValidityDays
    */
  if (updateUserPoolConfig.AdminCreateUserConfig && updateUserPoolConfig.AdminCreateUserConfig.UnusedAccountValidityDays) {
    delete updateUserPoolConfig.AdminCreateUserConfig.UnusedAccountValidityDays;
  }
  lambdaConfig.forEach((lambda) => (config[\`\${lambda.triggerType}\`] = lambda.lambdaFunctionArn));
  if (event.RequestType === 'Delete') {
    updateUserPoolConfig.LambdaConfig = {};
    console.log(\`\${event.RequestType}:\`, JSON.stringify(updateUserPoolConfig));
    const result = await cognitoClient.send(new UpdateUserPoolCommand(updateUserPoolConfig));
    console.log(\`delete response data \${JSON.stringify(result)}\`);
  } else if (event.RequestType === 'Update' || event.RequestType === 'Create') {
    updateUserPoolConfig.LambdaConfig = config;
    const result = await cognitoClient.send(new UpdateUserPoolCommand(updateUserPoolConfig));
    console.log(\`createOrUpdate response data \${JSON.stringify(result)}\`);
  }
}
",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "authTriggerFnServiceRole08093B67",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "authTriggerFnServiceRole08093B67": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "authTriggerFnServiceRoleDefaultPolicyEC9285A8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "cognito-idp:DescribeUserPool",
                "cognito-idp:UpdateUserPool",
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "userpoolArn",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "authTriggerFnServiceRoleDefaultPolicyEC9285A8",
        "Roles": [
          {
            "Ref": "authTriggerFnServiceRole08093B67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
